examples/notebooks/expression-tree.ipynb

```diff
@@ -0,0 +1,228 @@
+   "metadata": {},
+   "source": [
+    "![](expression_tree2.png)\n",
+    "\n",
+    "\n",
+    "...and evaluate this expression, which will again give 1."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "1"
+      ]
+     },
+     "execution_count": 4,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "diff_wrt_equation.evaluate(1, np.array([2]))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## The PyBaMM Pipeline\n",
+    "\n",
+    "Proposing, parameter setting and discretising a model in PyBaMM is a pipeline process, consisting of the following steps:\n",
+    "\n",
+    "1. The model is proposed, consisting of equations representing the right-hand-side of an ordinary differential equation (ODE), and/or algebraic equations for a differential algebraic equation (DAE), and also associated boundary condition equations\n",
+    "2. The parameters present in the model are replaced by actual scalar values from a parameter file, using the [`pybamm.ParamterValues`](https://pybamm.readthedocs.io/en/latest/source/parameters/parameter_values.html) class\n",
+    "3. The equations in the model are discretised onto a mesh, any spatial gradients are replaced with linear algebra expressions and the variables of the model are replaced with state vector slices. This is done using the [`pybamm.Discretisation`](https://pybamm.readthedocs.io/en/latest/source/discretisations/discretisation.html) class.\n",
+    "\n",
+    "## Stage 1 - Symbolic Expression Trees\n",
+    "\n",
+    "At each stage, the expression tree consists of certain types of nodes. In the first stage, the model is first proposed using [`pybamm.Parameter`](https://pybamm.readthedocs.io/en/latest/source/expression_tree/parameter.html), [`pybamm.Variable`](https://pybamm.readthedocs.io/en/latest/source/expression_tree/variable.html), and other [unary](https://pybamm.readthedocs.io/en/latest/source/expression_tree/unary_operator.html) and [binary](https://pybamm.readthedocs.io/en/latest/source/expression_tree/binary_operator.html) operators (which also includes spatial operators such as [`pybamm.Gradient`](https://pybamm.readthedocs.io/en/latest/source/expression_tree/unary_operator.html#pybamm.Gradient) and [`pybamm.Divergence`](https://pybamm.readthedocs.io/en/latest/source/expression_tree/unary_operator.html#pybamm.Divergence)). For example, the right hand side of the equation\n",
+    "\n",
+    "$$\\frac{d c}{dt} = D \\nabla \\cdot \\nabla c$$\n",
+    "\n",
+    "can be constructed as an expression tree like so:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "D = pybamm.Parameter('D')\n",
+    "c = pybamm.Variable('c', domain=['negative electrode'])\n",
+    "\n",
+    "rhs_c = D * pybamm.div(pybamm.grad(c))\n",
```

Comment:
It might be clearer to call this `dcdt` instead of `rhs_c`

Reply:
agreed